service/history/replication/ack_manager.go

```diff
@@ -503,3 +471,99 @@ func (p *ackMgrImpl) clusterToReaderID(
 
 	return readerID, nil
 }
+
+// TODO split the ack manager into 2 components
+
+func (p *ackMgrImpl) ConvertTask(
+	ctx context.Context,
+	task tasks.Task,
+) (*replicationspb.ReplicationTask, error) {
+	switch task := task.(type) {
+	case *tasks.SyncActivityTask:
+		return convertActivityStateReplicationTask(
+			ctx,
+			task,
+			p.workflowCache,
+		)
+	case *tasks.SyncWorkflowStateTask:
+		return convertWorkflowStateReplicationTask(
+			ctx,
+			task,
+			p.workflowCache,
+		)
+	case *tasks.HistoryReplicationTask:
+		return convertHistoryReplicationTask(
+			ctx,
+			task,
+			p.shard.GetShardID(),
+			p.workflowCache,
+			p.executionMgr,
+			p.logger,
+		)
+	default:
+		return nil, errUnknownReplicationTask
+	}
+}
+
+func (p *ackMgrImpl) SubscribeNotification() (<-chan struct{}, string) {
+	subscriberID := uuid.New().String()
+
+	p.subscriberLock.Lock()
+	defer p.subscriberLock.Unlock()
+
+	for {
+		if _, ok := p.subscribers[subscriberID]; !ok {
```

Comment:
How can this be not true? uuid collision?

Reply:
not ture == no collision => common case
true == uuid collision, retry